File: CPP/ValidParentheses.cpp
#include <iostream>
#include <stack>
#include <unordered_map>

bool isValid(const std::string& s) {
    std::stack<char> stack;
    std::unordered_map<char, char> bracket_map = {{')', '('}, {'}', '{'}, {']', '['}};
    
    for (char c : s) {
        if (bracket_map.find(c) != bracket_map.end()) {
            char top_element = stack.empty() ? '#' : stack.top();
            stack.pop();
            if (bracket_map[c] != top_element) {
                return false;
            }
        } else {
            stack.push(c);
        }
    }
    
    return stack.empty();
}

int main() {
    std::string s = "{[()]}";
    bool result = isValid(s);
    
    std::cout << "Valid Parentheses Result: " << (result ? "True" : "False") << std::endl;  // Output should be True
    
    return 0;
}

File: Python/Valid_Parentheses.py
def isValid(s: str) -> bool:
    # Dictionary to hold the mapping of opening and closing brackets
    bracket_map = {')': '(', '}': '{', ']': '['}
    # Initialize a stack to keep track of opening brackets
    stack = []
    
    # Iterate through each character in the string
    for char in s:
        # If it's a closing bracket
        if char in bracket_map:
            # Pop the topmost element from the stack if it's not empty
            # Otherwise, assign a dummy value
            top_element = stack.pop() if stack else '#'
            # Check if the popped element matches the corresponding opening bracket
            if bracket_map[char] != top_element:
                return False
        else:
            # If it's an opening bracket, push onto the stack
            stack.append(char)

    # In the end, the stack should be empty if all brackets are valid
    return not stack

# Example usage
if __name__ == "__main__":
    test_cases = [
        "()",         # True
        "()[]{}",     # True
        "(]",         # False
        "([)]",       # False
        "{[]}",       # True
    ]

    for test in test_cases:
        print(f"Input: {test}, Output: {isValid(test)}")
